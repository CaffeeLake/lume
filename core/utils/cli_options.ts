import { parseArgs } from "../../deps/cli.ts";
import type { DeepPartial } from "./object.ts";
import type { SiteOptions } from "../site.ts";

export function getOptionsFromCli(
  options: DeepPartial<SiteOptions>,
): DeepPartial<SiteOptions> {
  const cli = parseArgs(Deno.args, {
    string: ["src", "dest", "location", "port"],
    boolean: ["serve", "open"],
    alias: { dev: "d", serve: "s", port: "p", open: "o" },
    ["--"]: true,
  });

  if (cli.src) {
    options.src = cli.src;
  }

  if (cli.dest) {
    options.dest = cli.dest;
  }

  if (cli.location) {
    options.location = new URL(cli.location);
  } else if (cli.serve || cli._[0] === "cms") {
    options.location = new URL(`http://localhost:${cli.port || 3000}/`);
  }

  if (cli.port) {
    (options.server ||= {}).port = parseInt(cli.port);

    if (options.location) {
      options.location.port = cli.port;
    }
  }

  if (cli.open) {
    (options.server ||= {}).open = cli.open;
  }

  return options;
}
