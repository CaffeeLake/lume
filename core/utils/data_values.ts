import type { Data } from "../file.ts";

/**
 * Get the value of a page data
 * For example, if the value is "=title", it returns the value of the page data "title"
 * If the value is "$.title", it will return the value of the element with the selector ".title"
 */
export function getDataValue(data: Partial<Data>, value?: unknown) {
  // Get the value from the page data
  if (typeof value === "string") {
    if (value.startsWith("=")) {
      const key = value.slice(1);

      if (!key.includes(".")) {
        return data[key];
      }

      const keys = key.split(".");
      let val = data;
      for (const key of keys) {
        val = val?.[key];
      }
      return val;
    }

    if (value.startsWith("$")) {
      const document = data.page?.document;
      const query = value.slice(1);
      const element = document?.querySelector(query);
      return element?.innerHTML;
    }
  }

  if (typeof value === "function") {
    return value(data);
  }

  return value;
}
