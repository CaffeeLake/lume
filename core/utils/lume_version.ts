import { cyan, green, red } from "../../deps/colors.ts";

/** Return the latest stable version from the deno.land/x repository */
export async function getLatestVersion(): Promise<string> {
  const response = await fetch("https://cdn.deno.land/lume/meta/versions.json");
  const versions = await response.json();
  return versions.latest;
}

/** Return the hash of the latest commit from the GitHub repository */
export async function getLatestDevelopmentVersion(): Promise<string> {
  const response = await fetch(
    `https://api.github.com/repos/lumeland/lume/commits/main`,
  );
  const commits = await response.json();
  return commits.sha;
}

/** Return the current installed version */
export function getCurrentVersion(
  url = new URL(import.meta.resolve("../")),
): string {
  const { pathname } = url;
  return pathname.match(/@([^/]+)/)?.[1] ?? `local (${pathname})`;
}

/** Check the compatibility with the current Deno version */
export function checkDenoSupport(): void {
  const minimum = "1.38.0";
  const current = Deno.version.deno;

  if (current < minimum) {
    console.log("----------------------------------------");
    console.error(red("Your Deno version is not compatible with Lume"));
    console.log(`Lume needs Deno ${green(minimum)} or greater`);
    console.log(`Your current version is ${red(current)}`);
    console.log(`Run ${cyan("deno upgrade")} and try again`);
    console.log("----------------------------------------");
    Deno.exit(1);
  }
}
