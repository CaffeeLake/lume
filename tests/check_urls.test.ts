import { assertSiteSnapshot, build, getSite } from "./utils.ts";
import checkUrls from "../plugins/check_urls.ts";

Deno.test("check_urls plugin", async (t) => {
  const site = getSite({
    src: "check_urls",
  });

  site.use(checkUrls({
    output: "check_urls.json",
  }));

  await build(site);
  await assertSiteSnapshot(t, site);
});

Deno.test("check_urls plugin (strict mode)", async (t) => {
  const site = getSite({
    src: "check_urls",
  });

  site.use(checkUrls({
    output: "check_urls.json",
    strict: true,
  }));

  await build(site);
  await assertSiteSnapshot(t, site);
});
