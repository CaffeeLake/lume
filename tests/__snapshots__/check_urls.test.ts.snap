export const snapshot = {};

snapshot[`check_urls plugin 1`] = `
{
  formats: [
    {
      engines: 0,
      ext: ".page.toml",
      loader: [AsyncFunction: toml],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".page.ts",
      loader: [AsyncFunction: module],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".page.js",
      loader: [AsyncFunction: module],
      pageType: "page",
    },
    {
      engines: 0,
      ext: ".page.jsonc",
      loader: [AsyncFunction: json],
      pageType: "page",
    },
    {
      engines: 0,
      ext: ".page.json",
      loader: [AsyncFunction: json],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".json",
      loader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".jsonc",
      loader: [AsyncFunction: json],
    },
    {
      engines: 1,
      ext: ".md",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".markdown",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".js",
      loader: [AsyncFunction: module],
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".ts",
      loader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".vento",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".vto",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: toml],
      engines: 0,
      ext: ".toml",
      loader: [AsyncFunction: toml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yaml",
      loader: [AsyncFunction: yaml],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yml",
      loader: [AsyncFunction: yaml],
      pageType: "page",
    },
  ],
  src: [
    "/",
    "/index.md",
    "/page1.md",
    "/page4.md",
  ],
}
`;

snapshot[`check_urls plugin 2`] = `[]`;

snapshot[`check_urls plugin 3`] = `
[
  {
    content: '{
  "/page2/": [
    "/"
  ],
  "/page2": [
    "/"
  ]
}',
    data: {
      basename: "check_urls",
      content: '{
  "/page2/": [
    "/"
  ],
  "/page2": [
    "/"
  ]
}',
      page: [
        "src",
        "data",
        "asset",
      ],
      url: "check_urls.json",
    },
    src: {
      asset: true,
      ext: "",
      path: "",
      remote: undefined,
    },
  },
  {
    content: '<!DOCTYPE html>
<html><head></head><body><p><a href="./page1/">Relative link to existing page</a>
<a href="/page1">Link to existing page without trailin slash</a></p>
<p><a href="/page2/">Link to non existing page</a>
<a href="./page2">Relative link to non existing page without trailin slash</a></p>
<p><a href="/page3/">Link to redirected page</a>
<a href="./page3">Relative link to redirected page without trailin slash</a></p>
</body></html>',
    data: {
      basename: "index",
      children: '<p><a href="./page1/">Relative link to existing page</a>
<a href="/page1">Link to existing page without trailin slash</a></p>
<p><a href="/page2/">Link to non existing page</a>
<a href="./page2">Relative link to non existing page without trailin slash</a></p>
<p><a href="/page3/">Link to redirected page</a>
<a href="./page3">Relative link to redirected page without trailin slash</a></p>
',
      content: "[Relative link to existing page](./page1/)
[Link to existing page without trailin slash](/page1)

[Link to non existing page](/page2/)
[Relative link to non existing page without trailin slash](./page2)

[Link to redirected page](/page3/)
[Relative link to redirected page without trailin slash](./page3)
",
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/index",
      remote: undefined,
    },
  },
  {
    content: "<!DOCTYPE html>
<html><head></head><body><h1>Page 1</h1>
</body></html>",
    data: {
      basename: "page1",
      children: "<h1>Page 1</h1>
",
      content: "# Page 1

",
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/page1/",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/page1",
      remote: undefined,
    },
  },
  {
    content: "<!DOCTYPE html>
<html><head></head><body><h1>Page 3 is now 4</h1>
</body></html>",
    data: {
      basename: "page4",
      children: "<h1>Page 3 is now 4</h1>
",
      content: "# Page 3 is now 4
",
      date: [],
      mergedKeys: [
        "tags",
      ],
      oldUrl: "/page3/",
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/page4/",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/page4",
      remote: undefined,
    },
  },
]
`;

snapshot[`check_urls plugin (strict mode) 1`] = `
{
  formats: [
    {
      engines: 0,
      ext: ".page.toml",
      loader: [AsyncFunction: toml],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".page.ts",
      loader: [AsyncFunction: module],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".page.js",
      loader: [AsyncFunction: module],
      pageType: "page",
    },
    {
      engines: 0,
      ext: ".page.jsonc",
      loader: [AsyncFunction: json],
      pageType: "page",
    },
    {
      engines: 0,
      ext: ".page.json",
      loader: [AsyncFunction: json],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".json",
      loader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".jsonc",
      loader: [AsyncFunction: json],
    },
    {
      engines: 1,
      ext: ".md",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".markdown",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".js",
      loader: [AsyncFunction: module],
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".ts",
      loader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".vento",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".vto",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: toml],
      engines: 0,
      ext: ".toml",
      loader: [AsyncFunction: toml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yaml",
      loader: [AsyncFunction: yaml],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yml",
      loader: [AsyncFunction: yaml],
      pageType: "page",
    },
  ],
  src: [
    "/",
    "/index.md",
    "/page1.md",
    "/page4.md",
  ],
}
`;

snapshot[`check_urls plugin (strict mode) 2`] = `[]`;

snapshot[`check_urls plugin (strict mode) 3`] = `
[
  {
    content: '{
  "/page1": [
    "/"
  ],
  "/page2/": [
    "/"
  ],
  "/page2": [
    "/"
  ],
  "/page3/": [
    "/"
  ],
  "/page3": [
    "/"
  ]
}',
    data: {
      basename: "check_urls",
      content: '{
  "/page1": [
    "/"
  ],
  "/page2/": [
    "/"
  ],
  "/page2": [
    "/"
  ],
  "/page3/": [
    "/"
  ],
  "/page3": [
    "/"
  ]
}',
      page: [
        "src",
        "data",
        "asset",
      ],
      url: "check_urls.json",
    },
    src: {
      asset: true,
      ext: "",
      path: "",
      remote: undefined,
    },
  },
  {
    content: '<!DOCTYPE html>
<html><head></head><body><p><a href="./page1/">Relative link to existing page</a>
<a href="/page1">Link to existing page without trailin slash</a></p>
<p><a href="/page2/">Link to non existing page</a>
<a href="./page2">Relative link to non existing page without trailin slash</a></p>
<p><a href="/page3/">Link to redirected page</a>
<a href="./page3">Relative link to redirected page without trailin slash</a></p>
</body></html>',
    data: {
      basename: "index",
      children: '<p><a href="./page1/">Relative link to existing page</a>
<a href="/page1">Link to existing page without trailin slash</a></p>
<p><a href="/page2/">Link to non existing page</a>
<a href="./page2">Relative link to non existing page without trailin slash</a></p>
<p><a href="/page3/">Link to redirected page</a>
<a href="./page3">Relative link to redirected page without trailin slash</a></p>
',
      content: "[Relative link to existing page](./page1/)
[Link to existing page without trailin slash](/page1)

[Link to non existing page](/page2/)
[Relative link to non existing page without trailin slash](./page2)

[Link to redirected page](/page3/)
[Relative link to redirected page without trailin slash](./page3)
",
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/index",
      remote: undefined,
    },
  },
  {
    content: "<!DOCTYPE html>
<html><head></head><body><h1>Page 1</h1>
</body></html>",
    data: {
      basename: "page1",
      children: "<h1>Page 1</h1>
",
      content: "# Page 1

",
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/page1/",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/page1",
      remote: undefined,
    },
  },
  {
    content: "<!DOCTYPE html>
<html><head></head><body><h1>Page 3 is now 4</h1>
</body></html>",
    data: {
      basename: "page4",
      children: "<h1>Page 3 is now 4</h1>
",
      content: "# Page 3 is now 4
",
      date: [],
      mergedKeys: [
        "tags",
      ],
      oldUrl: "/page3/",
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/page4/",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/page4",
      remote: undefined,
    },
  },
]
`;
